icon_validator/rules/plugin_validators/version_bump_validator.py

```diff
@@ -0,0 +1,345 @@
+        if (SpecConstants.OUTPUT in remote) != (SpecConstants.OUTPUT in local):
+            raise ValidationException(f"Output section was added or removed without a major version increment."
+                                      f"{VersionBumpValidator.MAJOR_INSTRUCTIONS_STRING}")
+
+    def validate_all_types_exist(self, remote, local):
+        # checks if an entire type has been removed from spec
+        for type_key in remote[SpecConstants.TYPES]:
+            if type_key not in local[SpecConstants.TYPES]:
+                raise ValidationException(f"Type {type_key} has been removed without a major version increment."
+                                          f"{self.MAJOR_INSTRUCTIONS_STRING}")
+
+    def validate_no_types_changed(self, remote, local):
+        # verifies that types have not been changed between spec versions
+        if SpecConstants.TYPES in remote and SpecConstants.TYPES in local:
+            self.validate_all_types_exist(remote, local)
+            for type_key, type_value in remote[SpecConstants.TYPES].items():
+                for type_inner_key, type_inner_val in type_value.items():
+                    local_type_in = local[SpecConstants.TYPES][type_key]
+                    if type_inner_key not in local_type_in:
+                        raise ValidationException(f"Type {type_inner_key} removed from {type_key} without a major"
+                                                  f" version increment."
+                                                  f"{self.MAJOR_INSTRUCTIONS_STRING}")
+                    if type_inner_val[SpecConstants.TYPE] != local_type_in[type_inner_key][SpecConstants.TYPE]:
+                        raise ValidationException(f"Type {type_inner_key} changed in type {type_key} without a major"
+                                                  f" version increment."
+                                                  f"{VersionBumpValidator.MAJOR_INSTRUCTIONS_STRING}")
+
+                    if type_inner_val[SpecConstants.REQUIRED] != local_type_in[type_inner_key][SpecConstants.REQUIRED]:
+                        raise ValidationException(f"Type {type_inner_key} changed in type {type_key} without a major"
+                                                  f"version increment."
+                                                  f"{self.MAJOR_INSTRUCTIONS_STRING}")
+
+    def validate_no_inner_type_changes(self, remote, local):
+        self.abstract_validate_no_change(remote, local, SpecConstants.INPUT, SpecConstants.TYPE)
+        self.abstract_validate_no_change(remote, local, SpecConstants.OUTPUT, SpecConstants.TYPE)
+
+    def validate_no_titles_changed(self, remote, local):
+        self.abstract_validate_no_change(remote, local, SpecConstants.INPUT, SpecConstants.TITLE)
+        self.abstract_validate_no_change(remote, local, SpecConstants.OUTPUT, SpecConstants.TITLE)
+
+    def abstract_validate_no_change(self, remote, local, input_or_output, field):
```

Comment:
Type hinting here would help for readability. This is a string but corresponds to only one of two possible values.

Reply:
type hinting added pretty much everywhere now